RavenDB-17055 : Sharded ETL

aviv86 opened the pull request:
https://issues.hibernatingrhinos.com/issue/RavenDB-17055

garayx committed:
RavenDB-13113: delete sharded database values

aviv86 committed:
RavenDB-17055 : initial work : create sharded endpoints, basic tests

aviv86 committed:
RavenDB-17055 : sharded etl processes should use the same EtlProccessState

aviv86 committed:
RavenDB-17055 : etl failover tests

aviv86 committed:
RavenDB-17055 : handle sharded database on RemoveNodeFromCluster

aviv86 committed:
RavenDB-17055 : add CreateShardedDatabaseInCluster to test base

aviv86 committed:
RavenDB-17055 : initial work on sharded GetOngoingTaskInfo

ayende commented on src/Raven.Server/Documents/ETL/EtlProcess.cs:
```diff
@@ -675,7 +678,7 @@ public void Run()
                     var state  = LastProcessState = GetProcessState(Database, Configuration.Name, Transformation.Name);
 
-                    var loadLastProcessedEtag = state.GetLastProcessedEtagForNode(_serverStore.NodeTag);
+                    var loadLastProcessedEtag = state.GetLastProcessedEtag(Database.DbBase64Id);
```
Here, shouldn't we also try the `GetLastProcessedEtagForNode` if this returns `0` ? In case we have legacy data?

aviv86 replied:
fixed

ayende commented on src/Raven.Server/Documents/ShardedHandlers/ShardedOngoingTasks.cs:
```diff
@@ -0,0 +1,549 @@
+{
+    public class ShardedOngoingTasksHandler : ShardedRequestHandler
+    {
+        [RavenShardedAction("/databases/*/admin/connection-strings", "PUT")]
```
A lot of this looks exactly like the non sharded version, can we delagate that by calling the `OngoingTasksHandler`, instead?

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/ShardedHandlers/ShardedOngoingTasks.cs:
```diff
@@ -0,0 +1,549 @@
+                            throw new BadRequestException("Getting task info for " + OngoingTaskType.PullReplicationAsHub + " is not supported");
+
+                        case OngoingTaskType.SqlEtl:
+/*
```
why all the commented code?

aviv86 replied:
removed

ayende commented on src/Raven.Server/ServerWide/ClusterStateMachine.cs:
```diff
@@ -1166,24 +1165,44 @@ private void RemoveNodeFromCluster(ClusterOperationContext context, BlittableJso
+                        else
+                        {
+                            bool emptyShardTypology = true;
+                            foreach (var shardTopology in record.Shards)
```
I think this is problematic. 
What happens when we have:

* A,B,C
* Database with a shard on each?
* We now remove the node...
* That will delete *part* of the database, however, no?


Looks like that is something that we should guard against.

aviv86 replied:
yes, we will handle that in a separate ticket 
https://issues.hibernatingrhinos.com/issue/RavenDB-17161

aviv86 committed:
RavenDB-17055 : ETL process state should keep track of last processed etag per DbId

aviv86 committed:
RavenDB-17055 : add more tests + cleanup

aviv86 committed:
RavenDB-17055 : ShardedOngoingTasksHandler : implement missing sharded endpoints - delete ongoing task, reset etl, toggle task state, remove connection string, get connection string

aviv86 committed:
RavenDB-17055 : changes and refactoring :
- refactor some methods in OngoingTasksHandler, so that we can use them from ShardedOngoingTasksHandler and avoid code duplication
- sharded GetOngoingTaskInfo should return info on specific sharded etl task (e.g. 'foo$1') instead of aggregated info of all shards

aviv86 committed:
RavenDB-17055 : fixes after rebase

aviv86 committed:
RavenDB-17055 : fix flaky test

aviv86 committed:
RavenDB-17055 : add sharded elastic etl test

ppekrol commented on src/Raven.Server/Documents/ETL/EtlLoader.cs:
```diff
@@ -692,12 +692,15 @@ public void HandleDatabaseRecordChange(DatabaseRecord record)
         public void HandleDatabaseValueChanged(DatabaseRecord record)
         {
+            var dbName = record.DatabaseName;
+            ShardHelper.TryGetShardIndexAndDatabaseName(ref dbName);
```
can this return 'false' ? should we throw?

aviv86 replied:
no, this can return -1 if this is not a sharded name with `$` suffix, and then `dbName` is not changed 
this code is going to be called from non-shrded etl too so we shouldn't throw

ppekrol commented on src/Raven.Server/Documents/ETL/EtlProcess.cs:
```diff
@@ -89,10 +89,13 @@ public abstract class EtlProcess : IDisposable
         public static EtlProcessState GetProcessState(DocumentDatabase database, string configurationName, string transformationName)
         {
+            var databaseName = database.Name;
+            ShardHelper.TryGetShardIndexAndDatabaseName(ref databaseName);
```
can this return 'false' ? should we throw?

ppekrol commented on src/Raven.Server/ServerWide/ClusterStateMachine.cs:
```diff
@@ -1170,24 +1169,49 @@ private void RemoveNodeFromCluster(ClusterOperationContext context, BlittableJso
+                                continue;
+                            }
+                        }
+                        else
```
is this correct `else` ? we can go here when we have `IsSharded` but also when RelevantFor is false?

aviv86 replied:
fixed

ppekrol commented on src/Raven.Server/ServerWide/ClusterStateMachine.cs:
```diff
@@ -1456,24 +1480,42 @@ private static void CleanupDatabaseReplicationCertificate(ClusterOperationContex
+            else
+            {
+                // db can be idle when we are deleting it
+                serverStore?.IdleDatabases.TryRemove(record.DatabaseName, out _);
```
why is this needed?

ppekrol commented on src/Raven.Server/ServerWide/Commands/ETL/UpdateEtlProcessStateCommand.cs:
```diff
@@ -52,7 +53,10 @@ private UpdateEtlProcessStateCommand()
         {
-            return EtlProcessState.GenerateItemName(DatabaseName, ConfigurationName, TransformationName);
+            var databaseName = DatabaseName;
+            ShardHelper.TryGetShardIndexAndDatabaseName(ref databaseName);
```
can this fail?

arekpalinski commented on src/Raven.Server/Documents/ShardedHandlers/ShardedOngoingTasks.cs:
```diff
@@ -0,0 +1,135 @@
+                        case OngoingTaskType.Subscription:
+                        case OngoingTaskType.PullReplicationAsSink:
+                        case OngoingTaskType.Backup:
+                            // todo
```
Minor tip that I found very useful - in addition to adding `// todo` I usually add my login e.g. `// todo arek` this way I can very easily find all my TODOs

arekpalinski commented on src/Raven.Server/ServerWide/ClusterStateMachine.cs:
```diff
@@ -1456,24 +1480,42 @@ private static void CleanupDatabaseReplicationCertificate(ClusterOperationContex
+            {
+                for (int i = 0; i < record.Shards.Length; i++)
+                {
+                    var shard = record.DatabaseName + "$" + i;
```
I believe that's something what we'll commonly do. Maybe it would be good to create some shard helper class?

arekpalinski commented:
In general. Looks good to me

aviv86 committed:
RavenDB-17055 : address PR comments

ayende commented on src/Raven.Server/Documents/ShardedHandlers/ShardedOngoingTasks.cs:
```diff
@@ -0,0 +1,139 @@
+
+                            var index = ShardHelper.TryGetShardIndexAndDatabaseName(ref name);
+                            if (index == -1)
+                                throw new ArgumentException($"ETL task '{name}' is sharded, you must specify the shard index, for example : '{name}$0'");
```
What is the meaning hear, for `name$0` ? What is the user expected to do, pass the value for the specific instance?
Also, given that we are just forwarding, do we need that much code here?

ayende commented on src/Raven.Server/Documents/Sharding/ShardedRequestHandler.cs:
```diff
@@ -51,5 +55,24 @@ public void AddHeaders<T>(ShardedBaseCommand<T> command, Headers header)
+
+        public override async Task WaitForIndexToBeApplied(TransactionOperationContext context, long index)
+        {
+            var dbs = ServerStore.DatabasesLandlord.TryGetOrCreateShardedResourcesStore(ShardedContext.DatabaseName).ToList();
```
What is the meaning here? What is this for?

ayende replied:
Pretty sure that this is conceptually wrong, because you may be needing to wait on a value on a _different node_

ayende commented on src/Raven.Server/Web/RequestHandler.cs:
```diff
@@ -778,6 +778,12 @@ protected virtual OperationCancelToken CreateOperationToken()
+
+        public virtual Task WaitForIndexToBeApplied(TransactionOperationContext context, long index)
+        {
+            return Task.CompletedTask;
```
why?
